### PR TITLE
Fix invalid byte sequence on EncodedString#split

### PR DESCRIPTION
**Pull request by bf4**
From:
- https://github.com/rspec/rspec-core/pull/1760
- via https://github.com/rspec/rspec-support/pull/134

### CODE REVIEW

**Comment by JonRowe** [comment]
So this makes me a little uncomfortable given you are assuming that the bad bytes can be scrubbed by converting to UTF-8 and back, if this is an incompatible encoding with UTF-8 then won't this destroy the original encoding? 

Also can you remove the whitespace changes from the PR

**Comment by bf4** [comment]
So, the whitespace changes look sorta terrible in the unified diff, but if you split it or look at it alone, you'll see I'm actually fixing inconsistent whitespace (and some spelling, but that's unrelated to your comment).

I'm going to push a better fix on top of this momentarily.

**Comment by bf4** [comment]
Weird, that passed in https://travis-ci.org/rspec/rspec-support/builds/49911552

**Comment by bf4** [comment]
If this looks good I"ll rebase down to one commit.

**Comment by bf4** [comment]
Ok, I think this is it.

**Comment by bf4** [comment]
I'm not sure what to do about the JRuby failure, if it's expected behavior.  It is introduced in this PR.

**Comment by JonRowe** [comment]
screen shot 2015-02-08 at 21 35 56

**Comment by bf4** [comment]
I mean, it's not totally a failure, just a different behavior. The `é\x80` results in `\xC3\xA9\x80` Where `\xCE\xA9` is equivalent to the `é`.  If I remove the egrave, then the failure really is just that JRuby is leaving the `\x80` / `128.chr` alone for some reason. Still, probably a bug.

**Comment by bf4** [comment]
```bash
LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 \
   rvm ruby-1.9.2-p330,ruby-1.9.3-p551,ruby-2.0.0-p598,ruby-2.1.5,ruby-2.2.0,jruby-1.7.18,rbx-2.2.2 do \
   ruby -e 'p [RUBY_VERSION, RUBY_ENGINE, Encoding.default_external, __ENCODING__] << "\x80".force_encoding("US-ASCII").chars.map{|char| char.valid_encoding? ? char : "?" }.join.encode("UTF-8")'

 for version in 1.9 2.0; do \
   export JRUBY_OPTS="-Xcompat.version=${version}" ; \
   LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 \
   rvm jruby-1.7.18 do \
   ruby -e 'p [RUBY_VERSION, RUBY_ENGINE, Encoding.default_external, __ENCODING__] << "\x80".force_encoding("US-ASCII").chars.map{|char| char.valid_encoding? ? char : "?" }.join.encode("UTF-8")';
 done
```

```plain
["1.9.2", "ruby", #<Encoding:UTF-8>, #<Encoding:UTF-8>, "?"]
["1.9.3", "ruby", #<Encoding:UTF-8>, #<Encoding:UTF-8>, "?"]
["2.0.0", "ruby", #<Encoding:UTF-8>, #<Encoding:UTF-8>, "?"]
["2.1.5", "ruby", #<Encoding:UTF-8>, #<Encoding:UTF-8>, "?"]
["2.2.0", "ruby", #<Encoding:UTF-8>, #<Encoding:UTF-8>, "?"]
["1.9.3", "jruby", #<Encoding:UTF-8>, #<Encoding:UTF-8>, "\x80"]
["2.1.0", "rbx", #<Encoding:UTF-8>, #<Encoding:UTF-8>, "?"]
["1.9.3", "jruby", #<Encoding:UTF-8>, #<Encoding:UTF-8>, "\x80"]
["2.0.0", "jruby", #<Encoding:UTF-8>, #<Encoding:UTF-8>, "\x80"]
```

**Comment by bf4** [comment]
@JonRowe @myronmarston Are y'all ok with this?

**Comment by JonRowe** [comment]
So the test does nothing on JRuby? I think it should be pending if thats the case, that way we can have one spec, use the `:pending` metadata and we'll be alerted when JRuby fixes the issue.

File: lib/rspec/support/encoded_string.rb [hunk 1]
```diff
@@ -3,16 +3,32 @@ module Support
...
+      UTF_8    = "UTF-8"
+      US_ASCII = "US-ASCII"
+      #
+      # In MRI 2.1 'invalid: :replace' changed to also replace an invalid byte sequence
+      # see https://github.com/ruby/ruby/blob/v2_1_0/NEWS#L176
+      # https://www.ruby-forum.com/topic/6861247
+      # https://twitter.com/nalsh/status/553413844685438976
+      #
+      # For example, given:
+      #  "\x80".force_encoding("Emacs-Mule").encode(:invalid => :replace).bytes.to_a
```
Comment (JonRowe) [inline lib/rspec/support/encoded_string.rb]: Given the rest of this file this line needs one more " " added at the start (so indented by two spaces not 1).

File: lib/rspec/support/encoded_string.rb [hunk 2]
```diff
@@ -54,7 +70,7 @@ def to_s
         #     vs "\x80".encode('UTF-8','ASCII-8BIT', undef: :replace, replace: '<undef>')
         #     # => '<undef>'
         #   Encoding::CompatibilityError
-        #    when Enconding.compatbile?(str1, str2) is false
+        #    when Encoding.compatibile?(str1, str2) is nil
```
Comment (JonRowe) [inline lib/rspec/support/encoded_string.rb]: Similarly you've unified this elsewhere so this needs indenting an additional space and then the text beneath indenting to 4 spaces.

File: lib/rspec/support/encoded_string.rb (hunk 1 above)
Comment (bf4) [inline lib/rspec/support/encoded_string.rb]: good catch

File: lib/rspec/support/encoded_string.rb [hunk 3]
```diff
@@ -80,25 +96,35 @@ def to_s
         #     e.g. "\x80".force_encoding('ASCII-8BIT').encode('Emacs-Mule')
         #
         # Raised by byte <-> char conversions
-        #  RangeError: out of char range
-        #   e.g. the UTF-16LE emoji: 128169.chr
+        #   RangeError: out of char range
+        #    e.g. the UTF-16LE emoji: 128169.chr
```
Comment (JonRowe) [inline lib/rspec/support/encoded_string.rb]: And this needs one additional space too

**Comment by JonRowe** [comment]
Sorry nitpicking over the whitespace again because it should either be all fixed or none :P

File: lib/rspec/support/encoded_string.rb (hunk 2 above)
Comment (bf4) [inline lib/rspec/support/encoded_string.rb]: I swear I thought I did that...

**Comment by bf4** [comment]
@JonRowe Not at all.  re: the JRuby spec, you think I should remove the 'JRUBY' commit and add ` :pending => RSpec::Support::Ruby.jruby?` with a comment link to the bug?

**Comment by JonRowe** [comment]
> you think I should remove the 'JRUBY' commit and add :pending => RSpec::Support::Ruby.jruby? with a comment link to the bug?

Yes.

**Comment by bf4** [comment]
Ok, I think I fixed the whitespace and it should be good. At some point though, I start expecting to find I missed something and I'm past that point :)

**Comment by JonRowe** [comment]
Picked to `3-2-maintenance`

**Comment by bf4** [comment]
great!  back to rspec-core now

**Pull request merged**